src/librustc_resolve/late/diagnostics.rs

```diff
@@ -482,10 +482,16 @@ impl<'a> LateResolutionVisitor<'a, '_, '_> {
 
         let mut bad_struct_syntax_suggestion = |def_id: DefId| {
             let (followed_by_brace, closing_brace) = self.followed_by_brace(span);
-            let mut suggested = false;
-            match source {
-                PathSource::Expr(Some(parent)) => {
-                    suggested = path_sep(err, &parent);
+            let pad = match self.r.field_names.get(&def_id) {
+                Some(fields) if fields.is_empty() => "",
+                _ => " ",
+            };
+
+            match &source {
```

Comment:
```suggestion
            match source {
```
(`PathSource` is small and `Copy`.)

Reply:
I made this change to be able to match the type on the `span` field of these two different patterns:

https://github.com/rust-lang/rust/pull/74211/files#diff-2f2c99a2d0f3967b759026f70c4313b8R517-R518

```rust
                PathSource::Expr(Some(Expr { span, kind: ExprKind::Call(_, _), .. }))
                | PathSource::TupleStruct(span) => {
```

Because the contents of the `Expr` variant is `Option<&Expr>` its `span` field is `&Span` while the `TupleStruct` holds a `Span`. I can't use `ref` on one and not the other because that is different bindings for the same field (a restriction that might be lifted but is still there). Borrowing `source` makes them both be `&Span` without any further changes.

Reply:
Reworked things slightly to remove even more slightly duplicated code and reduced the `match &source` to a much smaller expression reverting to the previous code for the larger `match`.